### PR TITLE
feat(hooks): add useRange

### PR DESCRIPTION
**Pull request by Haroenv**
**Summary**

implement useRange

**Result**

example implemented RangeInput with approximately the same code as existing React InstantSearch, with the value/start transformations done in InstantSearch as well: https://github.com/algolia/instantsearch.js/blob/dcbbd8804b4b6471d24820b42826b57388974c27/src/widgets/range-input/range-input.tsx#L144-L154

FX-443

### CODE REVIEW

File: examples/hooks/components/RangeInput.tsx [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+          refine([from, to]);
+        }}
+      >
+        <input
+          className={'ais-RangeInput-input ais-RangeInput-input--min'}
+          type="number"
+          min={min}
+          max={max}
+          value={from}
+          placeholder={min?.toString()}
```
Comment (dhayab) [inline examples/hooks/components/RangeInput.tsx:58]: Apparently CSB doesn't play well with optional chaining (same for max input).

File: packages/react-instantsearch-hooks/README.md [hunk 2]
```diff
@@ -468,6 +468,97 @@ function Hits(props) {
...
+};
+```
+
+</details>
+
+**Example**
+
+```jsx
+function Range(props) {
+  const { min, max } = useRange(props);
```
Comment (dhayab) [inline packages/react-instantsearch-hooks/README.md:554]:
```suggestion
  const { range: { min, max } } = useRange(props);
```

File: packages/react-instantsearch-hooks/src/useRange.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+import { useConnector } from './useConnector';
+
+import type {
+  RangeConnectorParams,
+  RangeWidgetDescription,
+} from 'instantsearch.js/es/connectors/range/connectRange';
+
+export type UseRangeProps = RangeConnectorParams;
+
+export function useRange(props?: UseRangeProps) {
```
Comment (dhayab) [inline packages/react-instantsearch-hooks/src/useRange.ts:12]: I think the `attribute` prop makes this a required argument.

**Review by dhayab — APPROVED** [review]

File: examples/hooks/App.tsx [hunk 4]
```diff
@@ -70,6 +71,9 @@ export function App() {
               showMore={true}
             />
           </Panel>
+          <Panel header="price">
```
Comment (tkrugg) [inline examples/hooks/App.tsx:74]:
```suggestion
          <Panel header="Price">
```

**Review by tkrugg — APPROVED** [review]

File: examples/hooks/components/RangeInput.tsx [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    min:
+      minValue !== -Infinity && minValue !== min
+        ? minValue
+        : controlledUndefined,
+    max:
+      maxValue !== Infinity && maxValue !== max
+        ? maxValue
+        : controlledUndefined,
+  };
+  const [{ from, to }, setState] = useState({
```
Comment (francoischalifour) [inline examples/hooks/components/RangeInput.tsx:28]: Naming suggestion, to narrow it down:

```suggestion
  const [{ from, to }, setRange] = useState({
```

File: examples/hooks/components/RangeInput.tsx [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+          max={max}
+          value={to}
+          placeholder={max ? max.toString() : ''}
+          disabled={!canRefine}
+          onChange={(e) =>
+            setState({ from, to: Number(e.currentTarget.value) })
+          }
+        />
+        <button className={'ais-RangeInput-submit'} type="submit">
+          Apply
```
Comment (francoischalifour) [inline examples/hooks/components/RangeInput.tsx:78]: In IS.js [we use the word "Go"](https://instantsearchjs.netlify.app/stories/?path=/story/refinements-currentrefinements--with-rangeinput). Should we use the same here?

File: examples/hooks/components/RangeInput.tsx [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+      )}
+    >
+      <form
+        className={'ais-RangeInput-form'}
+        onSubmit={(e) => {
+          e.preventDefault();
+          refine([from, to]);
+        }}
+      >
+        <input
```
Comment (francoischalifour) [inline examples/hooks/components/RangeInput.tsx:52]: We don't use the `step` property?

File: examples/hooks/components/RangeInput.tsx [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+      className={cx(
+        'ais-RangeInput',
+        !canRefine && 'ais-RangeInput-noRefinement',
+        props.className
+      )}
+    >
+      <form
+        className={'ais-RangeInput-form'}
+        onSubmit={(e) => {
+          e.preventDefault();
```
Comment (francoischalifour) [inline examples/hooks/components/RangeInput.tsx:48]: For explicitness and consistency? (same below)

```suggestion
        onSubmit={(event) => {
          event.preventDefault();
```

File: examples/hooks/components/RangeInput.tsx [hunk 9]
```diff
@@ -0,0 +1,83 @@
+import React, { useEffect, useState } from 'react';
+
+import { useRange, UseRangeProps } from 'react-instantsearch-hooks';
+import { cx } from '../cx';
+
+export type RangeProps = React.ComponentProps<'div'> & UseRangeProps;
+
+// if the default value is undefined, React considers the component uncontrolled initially, which we don't want 0 or NaN as the default value
+const controlledUndefined = '' as unknown as number;
```
Comment (francoischalifour) [inline examples/hooks/components/RangeInput.tsx:9]: Alternative name suggestion: `unsetRangeValue`

File: examples/hooks/components/RangeInput.tsx (hunk 7 above)
Comment (Haroenv) [reply examples/hooks/components/RangeInput.tsx:52]: can set it, I'm not really a fan on how precision gets turned into step in InstantSearch.js, seems like needless indirection

File: examples/hooks/components/RangeInput.tsx (hunk 6 above)
Comment (Haroenv) [reply examples/hooks/components/RangeInput.tsx:78]: copied from InstantSearch.css, same as InstantSearch.js is also good

**Pull request merged**